### PR TITLE
Do not use atoi

### PR DESCRIPTION
**Pull request by InterestedInTechAndCake**
atoi() doesn't provide a convenient way to check for error conditions/invalid values, hence converted the calls to use strtol() instead so that we can return early when invalid values are detected.

This is to fix issue https://github.com/CTSRD-CHERI/chericat/issues/22

### CODE REVIEW

File: src/mem_scan.c [hunk 1]
```diff
@@ -67,7 +67,12 @@
  */
 void scan_mem(sqlite3 *db, char* arg_pid) 
 {
-	int pid = atoi(arg_pid);
+	char *pEnd;
+	long int pid = strtol(arg_pid, &pEnd, 10);
```
Comment (brooksdavis) [inline src/mem_scan.c:71]: I think this string paring should be done in scan_mem's caller (main). That would let you avoid the printf churn.

File: src/db_process.c [hunk 2]
```diff
@@ -267,9 +267,33 @@ static int vm_info_query_callback(void *all_vm_info_ptr, int argc, char **argv,
...
+	// as expected from the vm_info_captured data structure. 
+	// If the check fails, it means the data is invalid. Without
+	// a more sophisticated logic in place to handle invalid data, 
+	// we will just quit the program and let users to examine 
+	// and fix the data before running this function again.
+	char *prot_pEnd;
+	long int prot = strtol(argv[3], &prot_pEnd, 10);
+	if (*prot_pEnd != '\0') {
+		errx(1, "Invalid kve_protection, it should be an integer: %s", argv[3]);
+	}
```
Comment (brooksdavis) [inline src/db_process.c:281]: I'd be tempted to roll this into a utility function to avoid this single use variables.

File: src/ptrace_utils.c [hunk 3]
```diff
@@ -133,8 +133,12 @@ void read_data(int pid, void *addr, void *vptr, int len)
 }
 
 void read_lwps(char *arg_pid) {
-	int pid = atoi(arg_pid);
-	
+        char *pEnd;
+        long int pid = strtol(arg_pid, &pEnd, 10);
```
Comment (brooksdavis) [inline src/ptrace_utils.c:137]: The same comment as scan_mem except that apparently nothing calls this?

Comment (InterestedInTechAndCake) [reply src/ptrace_utils.c:137]: It was implemented when I started looking at kernel threads and hence trying to see what ptrace could give us. I haven't worked further on this yet, will extend this piece of code when I get back to it.

File: src/db_process.c (hunk 2 above)
Comment (InterestedInTechAndCake) [reply src/db_process.c]: Addressed

File: src/mem_scan.c (hunk 1 above)
Comment (InterestedInTechAndCake) [reply src/mem_scan.c]: Addressed.

File: src/ptrace_utils.c (hunk 3 above)
Comment (InterestedInTechAndCake) [reply src/ptrace_utils.c]: Addressed and move the error handling to the caller (when it's implemented).

File: src/chericat.c [hunk 4]
```diff
@@ -145,7 +146,14 @@ main(int argc, char *argv[])
...
 		case 'd':
-			set_print_level(atoi(optarg));
+		{
+			char *pEnd;
+			long int debug_level = strtol(optarg, &pEnd, 10);
+			if (*pEnd != '\0' || debug_level < 0 || debug_level > 3) {
+				errx(1, "Debug level can only be 0, 1, 2 or 3, with 0 being debug off and 3 being the most verbose");
+			}
+			set_print_level(debug_level);
+		}
```
[the remark this answers is not in the document]
Comment (InterestedInTechAndCake) [reply src/chericat.c:164]: I am using the getopt API to parse the options and it doesn't support multiple letters, to support what you suggested I will need to implement new logic to do the parsing, unless you know there is a simpler way?

[the remark this answers is not in the document]
Comment (bsdjhb) [reply src/chericat.c:164]: Normally for something like this you use a counter and have the getopt case bump the counter, e.g.:

```
static int verbose = 0;

....

    while ((ch = getopt(argc, argv, "v")) != -1)
        switch (ch) {
        ....
        case 'v':
             verbose++;
             break;
        }
```

Then later you might have:

```
    if (verbose > 0)
       printf("A verbose message\n");
    if (verbose > 1)
       printf("A more verbose message\n");
```

If you invoked this program with '-vv' it gets treated as '-v -v' by getopt(3) and `verbose` ends up being 2.

[the remark this answers is not in the document]
Comment (psjm3) [reply src/chericat.c]: Thank you @bsdjhb! The trick works. I have now committed the changes, and updated the usage to reflect the new way to indicate debugging level.

File: src/chericat.c [hunk 5]
```diff
@@ -55,25 +55,30 @@
...
 			"Options:\n"
-			"     -d Determine the level of debugging messages to be printed. If omitted,\n"
-			"        the default is INFO level\n"
+			"     -d The number of 'd' provided determines the level of debugging messages\n"
+		        "        to be printed. If omitted, the default is NO_PRINT level, meaning no\n"
+			"        debugging output. If one 'd' (i.e. -d) is provided, the level is set\n"
+			"        to INFO. If two 'd' (i.e. -dd) is provided, the level is set to VERBOSE.\n"
+			"        The most verbose level is three 'd' (i.e. -ddd), which is the\n"
+			"        TROUBLESHOOTING level. Any more 'd' after three will be ignored.\n" 
 			"     -f Provide the database name to capture the data collected by chericat.\n"
```
Comment (brooksdavis) [inline src/chericat.c:70]: Arguably too much detail for usage. Naming the levels isn't particularly informative.

File: src/chericat.c [hunk 6]
```diff
@@ -54,25 +55,30 @@
 
 static void usage()
 {
-	fprintf(stderr, "Usage: chericat [-d <debug level>] [-f <database name>] [-p <pid>] [-v]\n\t[-c <binary name>]\n"
-			"     debug level - 0 = No output; 1 = INFO; 2 = VERBOSE; 3 = TROUBLESHOOT\n"
-			"     pid - pid of the target process for a snapshot of caps info\n"
+	fprintf(stderr, "Usage: chericat [-d*] [-f <database name>] [-p <pid>] [-v]\n\t[-c <binary name>]\n"
```
Comment (brooksdavis) [inline src/chericat.c:58]: Not user what `*` is supposed to mean. You usually just specify that -d can be passed and say that additional times increases debugging.

**Review by brooksdavis — COMMENTED** [review]
This patch would benefit form rebasing and squashing.

File: src/chericat.c (hunk 6 above)
Comment (psjm3) [reply src/chericat.c]: Fixed - removed the * (it was supposed to represent zero or more) and reduced the details on the usage message. Also squashed the previous 5 commits.

Thanks!

File: src/chericat.c [hunk 7]
```diff
@@ -54,25 +55,26 @@
...
+	fprintf(stderr, "Usage: chericat [-d] [-f <database name>] [-p <pid>] [-v]\n\t[-c <binary name>]\n"
+			"     pid           - pid of the target process for a snapshot of caps info\n"
 			"     database name - name of the database to store data captured by chericat\n"
+			"     binary name   - name of the binary file for which to show the\n"
+			"                     capabilities located, with corresponding symbols\n"
 			"Options:\n"
-			"     -d Determine the level of debugging messages to be printed. If omitted,\n"
-			"        the default is INFO level\n"
+			"     -d The number of 'd' (up to three ds) provided determines the level of debugging messages\n"
+		        "        to be printed, the higher the number of ds the more verbose the messages are.\n"
```
Comment (brooksdavis) [inline src/chericat.c:65]:
```suggestion
			"     -d Enable debugging output. Repeated -d's (up to 3) increase verbosity.\n" 
```
Previous text was also >80 columns.

File: src/chericat.c (hunk 6 above)
Comment (psjm3) [reply src/chericat.c]: Thanks for the suggestion, committed.

**Review by brooksdavis — APPROVED** [review]
I'd squash it all when merging and make sure the commit title says what the commit does not that it's squashed.

**Pull request merged**